docs on adding recognizers (#508)
microsoft/presidio: docs/analyzer/adding_recognizers.md

    This class has the following functions:

    i. load: load a model / resource to be used during recognition

    ```python
    def load(self)
    ```

    ii. analyze: The main function to be called for getting entities out of the new recognizer:

    ```python
    def analyze(self, text, entities, nlp_artifacts)
    ```

    Notes:
    1. Each recognizer has access to different NLP assets such as tokens, [lemmas](https://en.wikipedia.org/wiki/Lemma_(morphology)), and more.
    These are given through the `nlp_artifacts` parameter.
    Refer to the [source code](../../presidio-analyzer/presidio_analyzer/entity_recognizer.py) for more information.

    2. The `analyze` method should return a list of [RecognizerResult](../../presidio-analyzer/presidio_analyzer/recognizer_result.py).

2. Add it to the recognizer registry using `registry.add_recognizer(my_recognizer)`.

### Creating a remote recognizer

A remote recognizer is an `EntityRecognizer` object interacting with an external service. The external service could be a 3rd party PII detection service or a custom service deployed in parallel to Presidio.

> [Click here for an example implementation of a `RemoteRecognizer`](../samples/python/example_remote_recognizer.py).

Reviewer: ```suggestion
[Sample implementation of a `RemoteRecognizer`](../samples/python/example_remote_recognizer.py).
```

    Refer to the [source code](../../presidio-analyzer/presidio_analyzer/entity_recognizer.py) for more information.

    2. The `analyze` method should return a list of [RecognizerResult](../../presidio-analyzer/presidio_analyzer/recognizer_result.py).

2. Add it to the recognizer registry using `registry.add_recognizer(my_recognizer)`.

### Creating a remote recognizer

A remote recognizer is an `EntityRecognizer` object interacting with an external service. The external service could be a 3rd party PII detection service or a custom service deployed in parallel to Presidio.

<details>
  <summary>Click here for a reference implementation of a `RemoteRecognizer`</summary>

Here's an illustrative example of how a `RemoteRecognizer` should be implemented. In this example, an external PII detection service exposes two APIs: `detect` and `supported_entities`. The class implemented here, `MyRemoteRecognizer`, uses the `requests` package to call the external service via HTTP.

In this code snippet, we simulate the external PII detector by using the Presidio analyzer. In reality, we would adapt this code to fit the external PII detector we have in hand.

> 

Reviewer: this can be reaplaced with a link to the class py file, maybe iwth just a one liner on running it (the __main__ )

> # Supporting new types of PII entities

Reviewer 2: ```suggestion
# Supporting detection of new types of PII entities
```

> Presidio can be extended to support new types of PII entities, and to support additional languages.

Reviewer 3: support detection of new types of PII entities (also in the title)

Reviewer 2: added some github "suggestions" for this comment so it is just a matter of click to add if Omri agrees to this

> Presidio can be extended to support new types of PII entities, and to support additional languages.

Reviewer 2: ```suggestion
Presidio can be extended to support detection of new types of PII entities, and to support additional languages.
```

- [Introduction to recognizer development](#introduction-to-recognizer-development)
- [Types of recognizer classes in Presidio](#types-of-recognizer-classes-in-presidio)
- [Extending the analyzer for additional PII entities](#extending-the-analyzer-for-additional-pii-entities)
- [Extending to additional languages](#extending-to-additional-languages)

## Introduction to recognizer development

Entity recognizers are Python objects capable of detecting one or more entities in a specific language.
In order to extend Presidio's detection capabilities to new types of PII entities,
these `EntityRecognizer` objects should be added to the existing list of recognizers.

## Types of recognizer classes in Presidio

The following class diagram shows the different types of recognizer families Presidio contains.

![Recognizers class diagram](../assets/recognizers_class_diagram.png)

- The `EntityRecognizer` is an abstract class for all recognizers.
- The `RemoteRecognizer` is an abstract class for calling external PII detectors.
See more info [here](#creating-a-remote-recognizer).
- The `LocalRecognizer` is an abstract class for all recognizers running within the Presidio-analyzer process.

> - The `PatternRecognizer` is an class for supporting regex and deny-list based recognition logic,

Reviewer 3: perhaps (just a suggestion):
- Abstract class `LocalRecognizer` - implemented by all recognizers running within the Presidio-analyzer process. 
- Class `PatternRecognizer` -  extended by regex and deny-list based recognition logic. It includes validation (e.g., with checksum) and context support. See an example [here](#simple-example).

Author: I find the sentence "extended by regex and deny-list based recognition logic." a bit confusing. The `PatternRecognizer` is the class which enables the logic of regex and deny-lists, it is not extended by it. If the current isn't clear enough, do you have a suggestion for rephrasing?

Reviewer 3: Maybe contains regex and deny-list based.. ?

Author: This is still confusing. The pattern recognizer class doesn't contain the regex logic (which is what a specific PII recognizer would contain). It contains helper function for users to implement regex.

including validation (e.g., with checksum) and context support. See an example [here](#simple-example).

## Extending the analyzer for additional PII entities

> First, a class based on `EntityRecognizer` needs to be created.

Reviewer 2: Suggestion: another option, that might increase readability

1. Create a new class based on `EntityRecognizer`
2. Add the new recognizer to the recognizer registry so that the `AnalyzerEngine` can use the new recognizer during analysis.

# Set up analyzer with our updated recognizer registry
analyzer = AnalyzerEngine(registry=registry)

# Run with input text
text="His name is Mr. Jones"
results = analyzer.analyze(text=text, language="en")
print(results)

```

Alternatively, we can add the recognizer directly to the existing registry:

```python
from presidio_analyzer import AnalyzerEngine

analyzer = AnalyzerEngine()

analyzer.registry.add_recognizer(titles_recognizer)

results = analyzer.analyze(text=text,language="en")
print(results)
```

### Creating a new `EntityRecognizer` in code

To create a new recognizer via code:

1. Create a new Python class which implements [LocalRecognizer](../../presidio-analyzer/presidio_analyzer/local_recognizer.py).
(`LocalRecognizer` implements the base [EntityRecognizer](../../presidio-analyzer/presidio_analyzer/entity_recognizer.py) class)

    This class has the following functions:

    i. load: load a model / resource to be used during recognition

    ```python
    def load(self)
    ```

    ii. analyze: The main function to be called for getting entities out of the new recognizer:

    ```python
    def analyze(self, text, entities, nlp_artifacts)
    ```

    Notes:

>     1. Each recognizer has access to different NLP assets such as tokens, lemmas, and more.

Reviewer 3: maybe lemmatizations? for those who are not familiar with the term ;)

Author: Added link to wiki
